src/NServiceBus.Core/PropertyDictionary.cs

```diff
@@ -0,0 +1,173 @@
+﻿namespace NServiceBus
+{
+    using System.Collections;
+    using System.Collections.Generic;
+
+    sealed class PropertyDictionary : IReadOnlyDictionary<string, string>
```

Comment:
while the implementation looks valid to me, I'd strongly prefer to find a simpler solution and not relying on a custom collection.